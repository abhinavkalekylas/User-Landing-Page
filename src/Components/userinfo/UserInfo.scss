$text-color: black; 

.userinfo_container {
    margin-top: 1rem;

    .form, .serinfo_label {
        color: $text-color;
    }

    .input {
        display: flex;
        flex-direction: row;
    }

    .userinfo_label {
        font-size: 1.2rem;
        display: flex;
        justify-content: center;
        align-items: center;
        margin: 0.5rem;
    }

    .userinfo_btn, .reset_btn {
        height: 3rem;
        margin: 1rem 4rem;   
    }
}
