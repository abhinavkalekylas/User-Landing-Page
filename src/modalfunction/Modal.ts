export interface User {
  id: number;
  name: string;
  email: string;
  gender: GenderType;
  status: StatusType;
}

export enum GenderType {
  MALE = "male",
  FEMALE = "female",
}

export enum StatusType {
  ACTIVE = "active",
  INACTIVE = "inactive",
}

export const validateEmail = (email: string) => {
  const regex = /^[^\s@]+@[^\s@]+\.[^\s@]{2,}$/i;
  if (regex.test(email)) {
    return true;
  }
  return false;
};
