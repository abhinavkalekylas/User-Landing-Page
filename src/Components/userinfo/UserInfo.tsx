import React, { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { toast } from "react-toastify";
import { Form, FormGroup, Button } from "reactstrap";
import "./UserInfo.scss";
import {
  GenderType,
  StatusType,
  User,
  validateEmail,
} from "../../modalfunction/Modal";
import postUser from "../adduser/postUser";
import putUser from "../edituser/putUser";

const UserInfo = ({
  userData,
  operation,
  closeModal,
}: {
  userData: User;
  operation: string;
  closeModal: () => void;
}) => {
  const navigate = useNavigate();

  const [userForm, setUserForm] = useState(userData);
  const [error, setError] = useState(false);

  useEffect(() => {
    setUserForm(userData);
  }, []);

  const handleChange = (e: { target: { name: string; value: string } }) => {
    setUserForm({
      ...userForm,
      [e.target.name]: e.target.value,
    });
  };

  const handleForm = async (e: { preventDefault: () => void }) => {
    e.preventDefault();
    if (userForm.name === "" || userForm.email === "") {
      toast.warn("Please enter given field", {
        position: "top-right",
        autoClose: 1500,
        hideProgressBar: false,
        closeOnClick: true,
        pauseOnHover: true,
        draggable: true,
        progress: undefined,
      });
      return;
    } else if (!validateEmail(userForm.email)) {
      toast.warn("Invalid email entered", {
        position: "top-right",
        autoClose: 1500,
        hideProgressBar: false,
        closeOnClick: true,
        pauseOnHover: true,
        draggable: true,
        progress: undefined,
      });
      return;
    } else {
      try {
        operation === "add"
          ? await postUser(userForm)
          : await putUser(userForm);

        toast.success(`User ${operation}`, {
          position: "top-right",
          autoClose: 1500,
          hideProgressBar: false,
          closeOnClick: true,
          pauseOnHover: true,
          draggable: true,
          progress: undefined,
        });
        closeModal();
        setTimeout(() => {
          navigate(0);
        }, 1500);
      } catch (error) {
        setError(true);
        toast.error(`Error while doing ${operation} operation`, {
          position: "top-right",
          autoClose: 1500,
          hideProgressBar: false,
          closeOnClick: true,
          pauseOnHover: true,
          draggable: true,
          progress: undefined,
        });
        console.log("Error while editing user ", error);
      }
    }
  };

  const resetForm = () => {
    setUserForm(userData);
  };

  return (
    <div className="userinfo_container">
      {error ? (
        <h2>This page is under develop. We will sure give you an update</h2>
      ) : (
        <>
          <Form onSubmit={handleForm} className="form">
            <FormGroup>
              <div className="input input-group input-group-lg">
                <label className="userinfo_label">Name :</label>
                <input
                  className="username form-control my-2"
                  placeholder="Enter username here"
                  aria-label="Sizing example input"
                  aria-describedby="inputGroup-sizing-lg"
                  onChange={handleChange}
                  type="text"
                  name="name"
                  value={userForm.name}
                  autoComplete="off"
                  required
                />
              </div>
              <div className="input input-group input-group-lg">
                <label className="userinfo_label">Email :</label>
                <input
                  className="email form-control my-2"
                  placeholder="Enter email id here"
                  aria-label="Sizing example input"
                  aria-describedby="inputGroup-sizing-lg"
                  onChange={handleChange}
                  type="text"
                  name="email"
                  value={userForm.email}
                  autoComplete="off"
                  required
                />
              </div>
              <div className="input input-group input-group-lg">
                <label className="userinfo_label">Gender :</label>
                <select
                  name="gender"
                  value={userForm.gender}
                  onChange={handleChange}
                  className="gender form-select form-select-lg mb-3"
                  aria-label=".form-select-lg example"
                >
                  <option value={GenderType.MALE}>Male</option>
                  <option value={GenderType.FEMALE}>Female</option>
                </select>
              </div>

              <div className="input input-group input-group-lg">
                <label className="userinfo_label">Status :</label>
                <select
                  name="status"
                  value={userForm.status}
                  onChange={handleChange}
                  className="status form-select form-select-lg mb-3"
                  aria-label=".form-select-lg example"
                >
                  <option value={StatusType.ACTIVE}>Active</option>
                  <option value={StatusType.INACTIVE}>Inactive</option>
                </select>
              </div>
              <div className="d-flex">
                <Button className="userinfo_btn" color="success" type="submit">
                  {operation} User
                </Button>

                <Button
                  className="reset_btn"
                  color="primary"
                  onClick={resetForm}
                >
                  Reset
                </Button>
              </div>
            </FormGroup>
          </Form>
        </>
      )}
    </div>
  );
};

export default UserInfo;
