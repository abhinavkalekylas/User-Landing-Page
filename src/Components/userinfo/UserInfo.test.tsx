import { BrowserRouter as Router } from "react-router-dom";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { GenderType, StatusType, User } from "../../modalfunction/Modal";
import UserInfo from "./UserInfo";

const userData: User = {
  id: 0,
  name: "shivam sharma",
  email: "shivam@gmail.com",
  gender: GenderType.MALE,
  status: StatusType.ACTIVE,
};

let wrapper: any;
const props = {
  userData,
  operation: "add",
  closeModal: jest.fn(),
};
beforeEach(() => {
  wrapper = shallow(
    <Router>
      <UserInfo {...props} />
    </Router>
  );
});

// snapshot testing
fit("renders correctly UserInfo component", () => {
  expect(toJson(wrapper)).toMatchSnapshot();
});
