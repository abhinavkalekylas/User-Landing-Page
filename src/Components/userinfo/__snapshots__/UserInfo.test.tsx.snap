// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly UserInfo component 1`] = `
<Router
  location={
    Object {
      "hash": "",
      "key": "default",
      "pathname": "/",
      "search": "",
      "state": null,
    }
  }
  navigationType="POP"
  navigator={
    Object {
      "action": "POP",
      "back": [Function],
      "block": [Function],
      "createHref": [Function],
      "forward": [Function],
      "go": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <UserInfo
    closeModal={[MockFunction]}
    operation="add"
    userData={
      Object {
        "email": "shivam@gmail.com",
        "gender": "male",
        "id": 0,
        "name": "shivam sharma",
        "status": "active",
      }
    }
  />
</Router>
`;
